File: Leetcode/cpp/medium/874.cpp
/*
 * @lc app=leetcode.cn id=874 lang=cpp
 * @lcpr version=30204
 *
 * [874] 模拟行走机器人
 */


// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
const vector<pair<int, int>> DIRS = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        int x = 0, y = 0, d = 0;
        int ans = 0;
        
        set<pair<int, int>> obstacle_set;
        for (const auto& obstacle : obstacles) {
            obstacle_set.insert({obstacle[0], obstacle[1]});
        }
        for (const auto& command : commands) {
            if (command < 0) {
                d = (d + (command == -1 ? 1 : 3)) % 4;
            } else {
                for (int i = 0; i < command; i++) {
                    int nx = x + DIRS[d].first, ny = y + DIRS[d].second;
                    if (obstacle_set.count({nx, ny})) {
                        break;
                    }
                    x = nx, y = ny;
                    ans = max(ans, x * x + y * y);
                }
            }
        }
        return ans;
    }
};
// @lc code=end



/*
// @lcpr case=start
// [4,-1,3]\n[]\n
// @lcpr case=end

// @lcpr case=start
// [4,-1,4,-2,4]\n[[2,4]]\n
// @lcpr case=end

// @lcpr case=start
// [6,-1,-1,6]\n[]\n
// @lcpr case=end

 */

File: Leetcode/rust/medium/874.rs
/*
 * @lc app=leetcode id=874 lang=rust
 *
 * [874] Walking Robot Simulation
 */

// @lcpr-template-start
struct Solution;
// @lcpr-template-end
// @lc code=start
impl Solution {
    pub fn robot_sim(commands: Vec<i32>, obstacles: Vec<Vec<i32>>) -> i32 {
        use std::collections::HashSet;

        let dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        let (mut x, mut y): (i32, i32) = (0, 0);
        let mut d: usize = 0;
        let mut ans: i32 = 0;
        let obstacles: HashSet<(i32, i32)> = obstacles.into_iter().map(|p| (p[0], p[1])).collect();

        for command in commands {
            if command < 0 {
                if command == -1 {
                    d = (d + 1) % 4;
                } else {
                    d = (d + 3) % 4;
                }
            } else {
                for _ in 0..command as usize {
                    let nx = x + dirs[d].0;
                    let ny = y + dirs[d].1;
                    if obstacles.contains(&(nx, ny)) {
                        break;
                    }
                    (x, y) = (nx, ny);
                    ans = ans.max(x * x + y * y);
                }
            }
        }
        ans
    }
}
// @lc code=end
